src/trio/_core/_run.py

```diff
@@ -2712,6 +2730,11 @@ def unrolled_run(
 
                 next_send_fn = task._next_send_fn
                 next_send = task._next_send
+                run_with = (
+                    run_with_ki_protection_enabled
+                    if task._ki_protected
+                    else run_with_ki_protection_disabled
+                )
```

Comment:
I would prefer a different approach to task-level KI protection. This one adds function call overhead to every task step, as well as a traceback frame for every level in the task tree. Here's an alternative idea: use a contextvar to indicate task-level KI protection, and have `_run` tell the KIManager which code object corresponds to `unrolled_run`. If stack unwinding reaches that object, KI protection should be based on the contextvar. `Context.run` is written in C, so as long as the contextvar is set to enable KI protection when outside of any task, this approach should be just as safe as the one you're using.

Reply:
Ah that's neat, but I don't think a contextvar is necessary. I think a threadlocal should be good enough

Reply:
The reason to use a contextvar is that we are already atomically switching contexts when we enter and exit the task. If you use a thread-local, then you have a brief period where the task is not running but the KI protection is still as if the task were running. This must be managed carefully in order to avoid an unprotected KI being raised directly inside `unrolled_run`. The contextvar solution is more obviously correct IMO.

Reply:
A contextvar won't work as someone could do:

```python
context = contextvars.copy_context()

async def some_task():
    def unprotected():
        ...
    context.run(unprotected)
```

Reply:
I've pushed a new solution that inspects to see if current_task().coro.cr_frame is in the stack, and if so returns current_task()._ki_protected